massa-pos-exports/src/pos_final_state.rs

```diff
@@ -53,17 +58,97 @@ impl PoSFinalState {
         let init_seed = Hash::compute_from(initial_seed_string.as_bytes());
         let initial_seeds = vec![Hash::compute_from(init_seed.to_bytes()), init_seed];
 
+        let end_slot = Slot::new(0, config.thread_count.saturating_sub(1));
+
         Ok(Self {
-            config,
+            config: config.clone(),
             cycle_history: Default::default(),
             deferred_credits: DeferredCredits::default(),
             selector,
             initial_rolls,
             initial_seeds,
             initial_ledger_hash,
+            initial_cycle: end_slot.get_cycle(config.periods_per_cycle),
+            last_start_period: end_slot.period,
+        })
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    /// create a `PoSFinalState` from an existing snapshot
+    pub fn from_snapshot(
+        config: PoSConfig,
+        cycle_history: VecDeque<CycleInfo>,
+        deferred_credits: DeferredCredits,
+        initial_seed_string: &str,
+        initial_rolls_in_snapshot: BTreeMap<Address, u64>,
+        selector: Box<dyn SelectorController>,
+        initial_ledger_hash: Hash,
+        end_slot: Slot,
+    ) -> Result<Self, PosError> {
+        // Seeds used as the initial seeds for negative cycles (initial_cycle-2 and initial_cycle-1 respectively)
+        let init_seed = Hash::compute_from(initial_seed_string.as_bytes());
+        let initial_seeds = vec![Hash::compute_from(init_seed.to_bytes()), init_seed];
+
+        Ok(Self {
+            config: config.clone(),
+            cycle_history,
+            deferred_credits,
+            selector,
+            initial_rolls: initial_rolls_in_snapshot,
+            initial_seeds,
+            initial_ledger_hash,
+            initial_cycle: end_slot.get_cycle(config.periods_per_cycle),
+            last_start_period: end_slot.period,
         })
     }
 
+    /// Handles the deferred credits that were due during a downtime
+    pub fn update_deferred_credits_after_restart(
+        &mut self,
+        from_slot: Slot,
+        end_slot: Slot,
+    ) -> Result<(), PosError> {
+        let mut deferred_credits = self.deferred_credits.clone();
```

Comment:
there's a few clones in here. can it be done without by any chance?

if partial move of self is a problem, one possibility is:

```rust
let Self { 
    /// fields go here
} = self;
```
that way, you don't have to use `self.$FIELD` anymore, and helps get around partial moves

Reply:
Righto.